crates/astria-sequencer-relayer/src/relayer/mod.rs

```diff
@@ -0,0 +1,304 @@
+        block_stream.set_block_time(self.sequencer_poll_period);
+
+        let mut forward_once_free: Fuse<
+            BoxFuture<Result<(), tokio::sync::mpsc::error::SendError<SequencerBlock>>>,
+        > = Fuse::terminated();
+
+        self.state.set_ready();
+
+        let reason = loop {
+            select!(
+                biased;
+
+                res = &mut forward_once_free, if !forward_once_free.is_terminated() => {
+                    block_stream.resume();
+                    // XXX: exiting because submitter only returns an error after u32::MAX
+                    // retries, which is practically infinity.
+                    if res.is_err() {
+                        let err = res.wrap_err(
+                            "submitter exited unexpectly while trying to forward block"
+                        );
+                        break err;
+                    }
+                }
+
+                Some(res) = latest_height_stream.next() => {
+                    match res {
+                        Ok(height) => {
+                            self.state.set_latest_observed_sequencer_height(height.value());
+                            debug!(%height, "received latest height from sequencer");
+                            block_stream.set_latest_sequencer_height(height);
+                        }
+                        Err(error) => {
+                            self.state.set_sequencer_connected(false);
+                            warn!(
+                                %error,
+                                "failed fetching latest height from sequencer; waiting until next tick",
+                            );
+                        }
+                    }
+                }
+
+                Some((height, fetch_result)) = block_stream.next() => {
+                    let block = match fetch_result.wrap_err_with(||
+                        format!(
+                            "relayer ultimately failed fetching sequencer block at height {height}"
+                    )) {
+                        // XXX: exiting because the fetch in block_stream errors after u32::MAX
+                        // retries, which is practically infinity.
+                        Err(err) => break Err(err),
+                        Ok(block) => block,
+                    };
+                    self.state.set_latest_fetched_sequencer_height(height.value());
+                    if let Err(err) = self.forward_block_for_submission(
+                        height,
+                        block,
+                        &mut block_stream,
+                        submitter.clone(),
+                        &mut forward_once_free,
+                    ).wrap_err("submitter exited unexpectly while trying to forward block") {
+                        // XXX: exiting because submitter only returns an error after u32::MAX
+                        // retries, which is practically infinity.
+                        break Err(err);
```

Comment:
i don't see the retry in `forward_block_for_submission`?

Reply:
Yeah, I can see how that comment is confusing. I have changed the comment to talk about how there is no logic to restart the task to which we are forwarding the blocks.